### PR TITLE
Rewrite the BasiQueue using channels

### LINKED ISSUE #1325: Rewrite BasicQueue using channels
As a follow up to https://github.com/paritytech/substrate/issues/1168, and in the context of our discussion on "fearless concurrency" cc @andresilva @tomusdrw, I've set out to rewrite the import queue using channels. The main benefit is the removal of locks, making the concurrency much easier to work with. 

It can best be summed up by replacing the below:
```rust
/// Basic block import queue that is importing blocks sequentially in a separate thread,
/// with pluggable verification.
pub struct BasicQueue<B: BlockT, V: 'static + Verifier<B>> {
	handle: Mutex<Option<::std::thread::JoinHandle<()>>>,
	data: Arc<AsyncImportQueueData<B>>,
	verifier: Arc<V>,
	block_import: SharedBlockImport<B>,
}

/// Locks order: queue, queue_blocks, best_importing_number
struct AsyncImportQueueData<B: BlockT> {
	signal: Condvar,
	queue: Mutex<VecDeque<(BlockOrigin, Vec<BlockData<B>>)>>,
	queue_blocks: RwLock<HashSet<B::Hash>>,
	best_importing_number: RwLock<<<B as BlockT>::Header as HeaderT>::Number>,
	is_stopping: AtomicBool,
}
```

With this:

```rust
/// Interface to a basic block import queue that is importing blocks sequentially in a separate thread,
/// with pluggable verification.
#[derive(Clone)]
pub struct BasicQueue<B: BlockT> {
    sender: Sender<BlockImportMsg<B>>,
}

struct BlockImporter<B: BlockT> {
    port: Receiver<BlockImportMsg<B>>,
    result_port: Receiver<BlockImportWorkerMsg<B>>,
    worker_sender: Sender<BlockImportWorkerMsg<B>>,
    queue_blocks: HashSet<B::Hash>,
    best_importing_number: <<B as BlockT>::Header as HeaderT>::Number,
    link: Option<Box<dyn Link<B>>>,
}

struct BlockImportWorker<B: BlockT, V: Verifier<B>> {
    result_sender: Sender<BlockImportWorkerMsg<B>>,
    block_import: SharedBlockImport<B>,
    verifier: Arc<V>,
}
```

What enables the removal of those `Mutex` and `RwLock` is the fact that the only thing that is shared with other threads is a `Box<ImportQueue>`, and it's implemenation `BasicQueue`, which is just a wrapper around a sender. The actual "work" is done by the `BlockImporter` and `BlockImportWorker`, in their own thread, and cross-thread communication is done via messages sent on channels. 

The previous `Mutex<VecDeque<(BlockOrigin, Vec<BlockData<B>>)>>` is gone, since it is replaced by the channels shared between importer and worker.

We get two threads, a "worker" that does only the actual importing, and an "importer" that forwards blocks to be imported to the worker, and also handles other requests such as "status" and "is_importing". The "importer" is also the one that communicates back to network using `Link`. 

It is still a work in progress at, yet it basically works. There are some test failures(maybe this new setup causes slight changes in business logic that have to be addressed), and the main roadblock seems to be deadlocks on `ChainSync`, used both by the `BlockImporter` via `Link`, and by `Protocol` itself. So when a test does for example a `net.peer(0).sync.status().sync.state`, we get what seems to be a deadlock. 

I'm planning to address this deadlock by removing locks in `Protocol`, which is tracked at https://github.com/paritytech/substrate/issues/1326

### LINKED ISSUE #1482: Move justification import to the import queue
#1410 introduces the ability to request and import justifications from the network. These are currently imported synchronously as they're received from the network, and should instead be imported asynchronously in the import queue. 

https://github.com/paritytech/substrate/blob/ea3d06a9fc232be6c31afc73718a3ecca13fd6ce/core/network/src/protocol.rs#L420-L425

### PR DESCRIPTION
**Pull request by gterzian**
Fix #1325 
Fix #1482

### CODE REVIEW

**Comment by gterzian** [comment]
This is currently written on top of https://github.com/paritytech/substrate/pull/1282, which makes this PR look bigger than it is...

**Comment by gterzian** [comment]
This is, I think, currently blocked on https://github.com/paritytech/substrate/issues/1326(removing the locks in the Protocol and ChainSync code)

**Comment by gterzian** [comment]
Ok this one is ready for review. 

@andresilva please note that this also takes care of #1482 ...

**Comment by gterzian** [comment]
@arkpar Would you be interested in reviewing the first commit?

@andresilva I guess you would be best placed to review the second?

File: core/consensus/common/src/import_queue.rs [hunk 1]
```diff
@@ -70,20 +72,17 @@ pub trait Verifier<B: BlockT>: Send + Sync + Sized {
...
+pub trait ImportQueue<B: BlockT>: Send + Sync + ImportQueueClone<B> {
 	/// Start background work for the queue as necessary.
 	///
 	/// This is called automatically by the network service when synchronization
 	/// begins.
-	fn start<L>(&self, _link: L) -> Result<(), std::io::Error> where
-		Self: Sized,
-		L: 'static + Link<B>,
-	{
+	fn start(&self, _link: Box<Link<B>>) -> Result<(), std::io::Error> {
```
[the remark this answers is not in the document]
Comment (gterzian) [reply core/consensus/common/src/import_queue.rs]: Ok so I've tried creating the workers lazily when starting, the problem with that is that you end-up trading a single box for 2-3 additional stuff stored on the `BasicQueue`, since those things are passed-in when `new` is called, which is in a different place from where `start` is called. I could do something like storing a bunch of `Option` and then `take` the stuff out of them, I guess, and that feels quite complicated. 

After that, another problem I ran in was that what was previously done with passing `Link` around as a generic param in the form of `start<L: 'static + Link<B>>`, actually requires the `start` method on the `ImportQueue` trait to be generic, and we can't `Box` a trait with generic methods...

The reason why `ImportQueue` is passed around in `Box`, is because it's now a stateless wrapper around a sender, while previously it would be `Arc<ImportQueue>` since the struct that implemented `ImportQueue` also contained all the state related to the implementation( see https://github.com/paritytech/substrate/blob/c188a3dc1505167748121a8f6cb9e970ff47df91/core/consensus/common/src/import_queue.rs#L113). 

I'm not very familiar with the cost tradeoffs between `Arc` and `Box`, and my feeling is that the use of box makes sense, and we're gaining because we're not sharing state anymore. If that costs us an additional box for the `Link`, which previously was a generic param, it might be worth it. 

What do you think?

File: core/consensus/common/src/import_queue.rs [hunk 2]
```diff
@@ -97,227 +96,445 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
...
 
-			match queue_lock.pop_front() {
-				Some(new_blocks) => new_blocks,
-				None => break,
+			if result.is_err() {
+				self.best_importing_number = Zero::zero();
+				has_error = true;
 			}
+
+			let link = match self.link.as_ref() {
```
[the remark this answers is not in the document]
Comment (gterzian) [reply core/consensus/common/src/import_queue.rs]: Ok so the thing is that a lot of the tests are relying on this, and as this PR plugs the actual `BasicQueue` into them. I've added a `trace` in case this were to actually happen in production...

File: core/consensus/common/src/import_queue.rs [hunk 3]
```diff
@@ -97,227 +96,445 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
...
+	fn note_useless_and_restart_sync(&self, _who: Origin, _reason: &str) {}
 	/// Restart sync.
-	fn restart(&self) { }
+	fn restart(&self) {}
 }
 
 /// Block import successful result.
 #[derive(Debug, PartialEq)]
-pub enum BlockImportResult<H: ::std::fmt::Debug + PartialEq, N: ::std::fmt::Debug + PartialEq> {
+pub enum BlockImportResult<N: ::std::fmt::Debug + PartialEq> {
```
[the remark this answers is not in the document]
Comment (gterzian) [reply core/consensus/common/src/import_queue.rs]: I couldn't quite remember why I did this, but they're actually a reason for it :) 

In the previous code, the `import-thread` the "handling of results" and the "importing" where all done by the "import-thread". Now the worker does the importing, and the importer handles the result(and handles request such as "is importing"). 

Note that `BlockImportResult` is only used in the `Ok` case, in the `Err` case `BlockImportError` is used. The error doesn't contain the hash. This previously didn't matter be because all the logic happened in the same thread/function. For example in order to remove the hashes from the queue of block to be imported, they would be stored in a variable here

https://github.com/paritytech/substrate/blob/c188a3dc1505167748121a8f6cb9e970ff47df91/core/consensus/common/src/import_queue.rs#L275

and handled after the importing was done at

https://github.com/paritytech/substrate/blob/c188a3dc1505167748121a8f6cb9e970ff47df91/core/consensus/common/src/import_queue.rs#L288

So I actually removed the hash from `BlockImportResult`, but each "result" message that comes in from the worker now contains a tuple of (hash, result), which is necessary because we need the hash also when the result in an error...

**Comment by gterzian** [comment]
>I'm also worried about some of the `expects` when sending/receiving messages to/from the workers, what happens if one of the worker threads aborts? I think we can keep the synchronous interface since the heavy work will be done on the block import worker but maybe we should expose errors?


The idea is that since `ChainSync` is holding the `Box<ImportQueue`, the importer will not quit since that keeps a sender alive. `ChainSync` will itself [tell the ImportQueue to stop](https://github.com/paritytech/substrate/pull/1327/files#diff-0c805d3ef65d16a3738f8270c6726d38R583), and will [ignore any calls to `maintain_sync` after that](https://github.com/paritytech/substrate/pull/1327/files#diff-0c805d3ef65d16a3738f8270c6726d38R556). 

When sync stops, that means protocol stops, and protocol stops only when it's sender has been dropped, meaning the network service itself has dropped. 

https://github.com/paritytech/substrate/blob/9bb2125ec4d7470426813b03856550451f631f32/core/network/src/protocol.rs#L328 

I've reworded the expectations for clarity.

**Comment by gavofyork** [comment]
@arkpar @andresilva any further thoughts?

File: core/consensus/common/src/import_queue.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (andresilva) [reply core/consensus/common/src/import_queue.rs]: I'm also not very familiar with those cost tradeoffs, but I agree with your points. The only thing I don't like about the worker is really the exceptional "not started" state, I think an option would be to make it kind of a "builder pattern", so that `start` would return a new type (which just wraps the sender). But I'm not sure that would work and I don't have a strong opinion on this, I'm fine with logging the exceptional case.

File: core/consensus/common/src/import_queue.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (andresilva) [reply core/consensus/common/src/import_queue.rs]: Yeah I know the same behavior is being kept, I just meant from a semantic point of view. Also no strong opinion on this, fine as is.

**Comment by andresilva** [comment]
@gterzian There's still something I'm not understanding wrt that setup.

https://github.com/paritytech/substrate/blob/0af96432a7fcbd8fd3bf83bae65ad83f052f7ccb/core/consensus/common/src/import_queue.rs#L438-L447

This `sender` will be passed as `worker_sender` to the `BlockImporter`. If this thread panics then `port` will be dropped, and any send to the corresponding sender will fail. Such as this one:

https://github.com/paritytech/substrate/blob/0af96432a7fcbd8fd3bf83bae65ad83f052f7ccb/core/consensus/common/src/import_queue.rs#L420-L423

**Comment by arkpar** [comment]
Ideally we want the block import pipeline to call `verify` and `import_block` in parallel threads. Is that something that can be done easily with this design?

**Comment by arkpar** [comment]
Regarding the panics, we have a custom panic hook that exits the process in case of any panic, other than ones generated in the srml.

**Comment by gterzian** [comment]
> Ideally we want the block import pipeline to call `verify` and `import_block` in parallel threads. Is that something that can be done easily with this design?
 
Yes.

> Regarding the panics, we have a custom panic hook that exits the process in case of any panic, other than ones generated in the srml.

Ok, thanks for the info.

**Comment by gterzian** [comment]
Ok I fixed an error introduced in the rebase, and tested on charred-cherry(which seems to work on master again), which seemed to work fine...

File: core/consensus/aura/src/lib.rs [hunk 4]
```diff
@@ -615,7 +615,7 @@ impl<B: Block, C, E> Verifier<B> for AuraVerifier<C, E> where
 }
 
 /// The Aura import queue type.
-pub type AuraImportQueue<B, C, E> = BasicQueue<B, AuraVerifier<C, E>>;
+pub type AuraImportQueue<B> = BasicQueue<B>;
```
Comment (andresilva) [inline core/consensus/aura/src/lib.rs]: Maybe we can remove this type, since the `BasicQueue` type is now an opaque interface that will be used by everyone (regardless of underlying verifier).

File: core/consensus/common/src/import_queue.rs [hunk 5]
```diff
@@ -97,227 +96,448 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
...
-				justification_import.on_start(&link);
-			}
-			import_thread(block_import, link, qdata, verifier)
-		})?);
+impl<B: BlockT> ImportQueue<B> for BasicQueue<B> {
+	fn start(&self, link: Box<Link<B>>) -> Result<(), std::io::Error> {
+		let _ = self
+			.sender
+			.send(BlockImportMsg::Start(link))
+			.expect("1. self is holding a sender to the Importer, 2. Importer should handle messages while there are senders around");
```
Comment (andresilva) [inline core/consensus/common/src/import_queue.rs]: We usually write these proofs in the form:
"self is holding a sender to the Importer; Importer should handle messages while there are senders around; qed"

File: core/consensus/common/src/import_queue.rs [hunk 6]
```diff
@@ -97,227 +96,448 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
...
-			// We are holding the same lock that `stop` takes so here we either see that stop flag
-			// is active or wait for the signal. The latter one unlocks the mutex and this gives a chance
-			// to `stop` to generate the signal.
-			if qdata.is_stopping.load(Ordering::SeqCst) {
-				break;
+pub enum BlockImportWorkerMsg<B: BlockT> {
+	ImportBlocks(BlockOrigin, Vec<IncomingBlock<B>>),
+	Imported(
+		Vec<(
+			Result<BlockImportResult<<<B as BlockT>::Header as HeaderT>::Number>, BlockImportError>,
```
Comment (andresilva) [inline core/consensus/common/src/import_queue.rs]:
```suggestion
			Result<BlockImportResult<NumberFor<Block>>, BlockImportError>,
```

File: core/consensus/common/src/import_queue.rs [hunk 7]
```diff
@@ -97,227 +96,448 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
...
+impl<B: BlockT> ImportQueueClone<B> for BasicQueue<B> {
+	fn clone_box(&self) -> Box<ImportQueue<B>> {
+		Box::new(self.clone())
 	}
 }
 
-impl<B: BlockT> AsyncImportQueueData<B> {
-	/// Instantiate a new async import queue data.
-	pub fn new() -> Self {
+impl<B: BlockT> BasicQueue<B> {
```
Comment (andresilva) [inline core/consensus/common/src/import_queue.rs]: Can you add a general description here about how the queue works? e.g. how messages are dispatched to a worker, and blocks are imported on a different worker.

File: core/consensus/common/src/import_queue.rs [hunk 8]
```diff
@@ -97,227 +96,448 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
...
-				None => break,
+			if result.is_err() {
+				self.best_importing_number = Zero::zero();
+				has_error = true;
 			}
+
+			let link = match self.link.as_ref() {
+				Some(link) => link,
+				None => {
+					trace!(target:"sync", "Received import result for {} while import-queue has no link", hash);
```
Comment (andresilva) [inline core/consensus/common/src/import_queue.rs]:
```suggestion
					trace!(target: "sync", "Received import result for {} while import queue has no link", hash);
```

File: core/consensus/common/src/import_queue.rs [hunk 9]
```diff
@@ -97,227 +96,448 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
...
+					return true
```
Comment (andresilva) [inline core/consensus/common/src/import_queue.rs]:
```suggestion
					return true;
```

File: core/consensus/common/src/import_queue.rs [hunk 10]
```diff
@@ -97,227 +96,448 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
...
+			.expect("ImportQueueWorker thread spawning failed");
+		sender
+	}
+
+	fn import_a_batch_of_blocks(&self, origin: BlockOrigin, blocks: Vec<IncomingBlock<B>>) {
+		let count = blocks.len();
+		let mut imported = 0;
+
+		let blocks_range = match (
+			blocks
```
Comment (andresilva) [inline core/consensus/common/src/import_queue.rs]: This formatting is weird, it should be fine the keep the expression on one line like:
`blocks.first().and_then(|b| b.header.as_ref().map(|h| h.number())),`

File: core/consensus/common/src/import_queue.rs [hunk 11]
```diff
@@ -97,227 +96,448 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
...
+   					&*self.block_import,
+   					origin.clone(),
+   					block.clone(),
+   					self.verifier.clone(),
+   				)
+			};
+			let was_ok = import_result.is_ok();
+			results.push((import_result, block.hash));
+			if was_ok {
+				imported = imported + 1;
```
Comment (andresilva) [inline core/consensus/common/src/import_queue.rs]:
```suggestion
				imported += 1;
```

File: core/network/src/protocol.rs [hunk 12]
```diff
@@ -258,6 +258,8 @@ pub enum ProtocolMsg<B: BlockT, S: NetworkSpecialization<B>,> {
 	BlockImportedSync(B::Hash, NumberFor<B>),
 	/// Tell protocol to request justification for a block.
 	RequestJustification(B::Hash, NumberFor<B>),
+	/// Inform protocol whether a justification was successfully imported.
+	JustificationImportResut(B::Hash, NumberFor<B>, bool),
```
Comment (andresilva) [inline core/network/src/protocol.rs]:
```suggestion
	JustificationImportResult(B::Hash, NumberFor<B>, bool),
```

File: core/network/src/test/sync.rs [hunk 13]
```diff
@@ -53,11 +52,6 @@ fn sync_long_chain_works() {
 	let mut net = TestNet::new(2);
 	net.peer(1).push_blocks(500, false);
 	net.sync();
-	// Wait for peer 0 to import blocks received over the network.
```
Comment (andresilva) [inline core/network/src/test/sync.rs]: Why were these removed? (I'm not familiar with this test in particular)

File: core/consensus/common/src/import_queue.rs [hunk 14]
```diff
@@ -400,6 +400,9 @@ impl<B: BlockT> BlockImporter<B> {
 		let success = self.justification_import.as_ref().map(|justification_import| {
 			justification_import.import_justification(hash, number, justification).is_ok()
 		}).unwrap_or(false);
+		if !success {
+			debug!("Justification import failed for hash: {:?} number: {:?} coming from node: {:?}", hash, number, who);
```
Comment (andresilva) [inline core/consensus/common/src/import_queue.rs]: Can we move this to inside the `map` above so we can log the error itself?

File: core/consensus/aura/src/lib.rs (hunk 4 above)
Comment (gterzian) [reply core/consensus/aura/src/lib.rs]: The problem with this is that the `AuraImportQueue` is exported and used outside of the crate, while `BasicQueue` is private(and probably should be). Perhaps `AuraImportQueue` can be replaced by a `ImportQueue`, however I think that's going to require boxing it 😄 , and will also require some additional work since I would be very surprised if it didn't require changing some of the code using it. Perhaps we can look into this as a follow-up?

Comment (andresilva) [reply core/consensus/aura/src/lib.rs]: Didn't think about that. Sure that's reasonable. :+1:

File: core/network/src/test/sync.rs (hunk 13 above)
Comment (gterzian) [reply core/network/src/test/sync.rs]: I think it's a result of this change: https://github.com/paritytech/substrate/pull/1327/files/8113ac554afb9ed430b565e4569d9bc695c06a55#diff-d19df74b5b495aae3892f8f92a5f174aR213

It appears unnecessary to have the extra wait in the test...

File: core/consensus/common/src/import_queue.rs (hunk 14 above)
Comment (gterzian) [reply core/consensus/common/src/import_queue.rs]: yes indeed
